ChainSafe/gossamer
# feat(telemetry): send telemetry messages when GRANDPA receieves commit or vote messages

## Changes



- Send `afg.received_commit` when grandpa receives a commit message.
- Send `afg.received_precommit` or `afg.received_prevote` when grandpa
receives a vote message

## Tests




- Setup a local telemetry server using substrate-telemetry.

Use the branch `print-payload-type` from my fork `https://github.com/kishansagathiya/substrate-telemetry` (It print the message received, type of message and error. From what I have played around upstream doesn't do that.)

Run telementry backend (`telemetry_core` and `telemetry_shard`).

-  Run gossamer by setting telemetry  url to the local server. 

In order to run perform grandpa rounds, we will need to run a dev net. https://hackmd.io/@nZ-twauPRISEa6G9zg3XRw/B1RFO3uDt shows how to do that.
I ran it with 3 nodes. After running all the nodes, you should be able to see in `telemetry_shard` logs that we are receiving all 3 expected telemetry messages from time to time.


```

```

## Issues


- Closes #1840
- Closes #1839
- Closes #1838

## Primary Reviewer



- @noot
